mysys/my_atomic_writes.c

```diff
@@ -358,12 +376,68 @@ static my_bool sfx_has_atomic_write(File file, int page_size)
   }
   return 0;
 }
+
+static my_bool sfx_dev_could_disable_punch_hole(struct sfx_dev *dev, File file)
+{
+  int result = 0;
+
+  if (dev->disable_punch_hole == SFX_UNKNOWN_PUNCH_HOLE_YET)
+  {
+    int fd= open(dev->dev_name, 0);
+    if (fd < 0)
+    {
+      perror("open() failed!");
+      dev->disable_punch_hole= 0;                      /* Don't try again */
+      return FALSE;
+    }
+
+    /*
+      Ratio left-shifts 8 (multiplies 256) inside the ioctl;
+      will also add 1 to guarantee a round-up integer.
+    */
+    result= ioctl(fd, SFX_GET_SPACE_RATIO);
+    result+= 1;
+    dev->disable_punch_hole= (result >= (((double)SFX_DISABLE_PUNCH_HOLE_RATIO) * 256));
+  }
+
+  return dev->disable_punch_hole;
+}
+
+/**
+   Check if a file is on a ScaleFlux device and whether it is possible to
+   disable hole punch.
+   @param[in] file              OS file handle
+   @return TRUE                 Could disable hole punch
+
+   @notes
+   This is called only at first open of a file. In this case it's doesn't
+   matter so much that we loop over all cards
+*/
+
+static my_bool sfx_could_disable_punch_hole(File file)
+{
+  struct sfx_dev *dev;
+  struct stat stat_buff;
+
+  if (fstat(file, &stat_buff) < 0)
+  {
+    return 0;
+  }
+
+  for (dev = sfx_devices; dev->st_dev; dev++)
+  {
+    if (stat_buff.st_dev == dev->st_dev)
+      return sfx_dev_could_disable_punch_hole(dev, file);
```

Comment:
Also this was copied from somewhere else in this file. This would only work if the file system is created directly on the block device, without creating a partition table. Linux seems to allocate 16 numbers for each block device, so that there can be up to 15 partitions. We must ignore the partition number when comparing.